magic_items/routes.py

```diff
@@ -0,0 +1,304 @@
+from flask import Blueprint, render_template, g, flash, redirect, request
+from magic_items.services import (
+    MagicItemService,
+    ItemForm,
+    ItemFilterForm,
+    CollectionAddForm,
+    CollectionService,
+    ItemCollectionService,
+)
+from users.services import UserService
+
+magic_routes = Blueprint("magic_routes", __name__)
+
+
+def before_manage_magic_items(func):
+    def wrapper(*args, **kwargs):
+        if not g.user:
+            flash("Please login or signup to manage magic items", "warning")
+            return redirect("/login")
+        return func(*args, **kwargs)
+
+    wrapper.__name__ = func.__name__ + wrapper.__name__
+    return wrapper
+
+
+def before_manage_collections(func):
+    def wrapper(*args, **kwargs):
+        if not g.user:
+            flash("Please login or signup to manage collections", "warning")
+            return redirect("/login")
+        return func(*args, **kwargs)
+
+    wrapper.__name__ = func.__name__ + wrapper.__name__
+    return wrapper
+
+
+#########################################################################################################
+#########################################################################################################
+##
+## Magic Item Routes
+
+
+@magic_routes.route("/magic-items/new", methods=["GET", "POST"])
+@before_manage_magic_items
+def add_item():
+
+    form = ItemForm()
+
+    if form.validate_on_submit():
+        try:
+            item = MagicItemService.create(g.user.id, form)
+            flash("new item " + item.name + " added", "success")
+            return redirect(f"/magic-items/{item.id}")
+
+        except:
+            flash("error", "danger")
+            return render_template("magic_items/new_magic_item.html", form=form)
+
+    return render_template("magic_items/new_magic_item.html", form=form)
```

Comment:
If the form is not valid, we just return the same template again without telling the users what's missing.

Reply:
Currently errors are handled by an error loop on the form object on the template. Would it make more sense to handle this in the route?